storage/storage_image.go

```diff
@@ -90,309 +95,654 @@ func newImageSource(imageRef storageReference) (*storageImageSource, error) {
 	return image, nil
 }
 
-// newImageDestination sets us up to write a new image.
+// Reference returns the image reference that we used to find this image.  Since we're officially
+// unnamed, return a reference with no naming information.
+func (s storageUnnamedImageReader) Reference() types.ImageReference {
+	return newReference(s.transport, "", s.ID, nil, "", "")
+}
+
+// Close cleans up any resources we tied up while reading the image.
+func (s storageUnnamedImageReader) Close() error {
+	return nil
+}
+
+// GetBlob reads the data blob or filesystem layer which matches the digest and size, if given.
+func (s *storageUnnamedImageReader) GetBlob(info types.BlobInfo) (rc io.ReadCloser, n int64, err error) {
+	rc, n, _, err = s.getBlobAndLayerID(info, false)
+	return rc, n, err
+}
+
+// GetBlob reads the data blob or filesystem layer which matches the digest and size, if given, and returns
+// the layer ID for the layer, if it was a layer.
+func (s *storageUnnamedImageReader) getBlobAndLayerID(info types.BlobInfo, decompress bool) (rc io.ReadCloser, n int64, layerID string, err error) {
+	var layer storage.Layer
+	var diffOptions *storage.DiffOptions
+	// We need a valid digest value.
+	err = info.Digest.Validate()
+	if err != nil {
+		return nil, -1, "", err
+	}
+	// Check if the blob corresponds to a diff that was used to initialize any layers, either
+	// before or after decompression, since we don't care.
+	uncompressedLayers, err := s.transport.store.LayersByUncompressedDigest(info.Digest)
+	compressedLayers, err := s.transport.store.LayersByCompressedDigest(info.Digest)
```

Comment:
Yes, I suppose that though `storageUnnamedImageReader` is not dealing with an updated manifest, it's not going to get called to read blobs using the digests from the not-yet-updated manifest that it would return.  Reworking it to only look for layers using the index of uncompressed digests.